slm_porosity/heat_transfer.f90

```diff
@@ -737,54 +747,114 @@ FUNCTION porosity_term ()
+        Dconductivity_solid, Dconductivity_liquid, conductivity_fusion)
+    ELSE
+      conductivity = three_parameter_connection(temperature, liquid_fraction, &
+        Dconductivity_solid, Dconductivity_liquid, conductivity_fusion, .TRUE.)
+    END IF
   END FUNCTION conductivity
 
-  PURE FUNCTION capacity (temperature)
+  FUNCTION capacity (temperature, liquid_fraction, is_D)
     IMPLICIT NONE
-    REAL (pr), INTENT(IN) :: temperature(:)
+    REAL (pr), INTENT(IN) ::  temperature(:), liquid_fraction(:)
     REAL (pr) :: capacity(SIZE(temperature))
-    capacity = 1.0_pr + 2*capacity_der*temperature
+    LOGICAL, OPTIONAL, INTENT(IN) :: is_D
+
+    IF (.NOT.PRESENT(is_D).OR.(.NOT.is_D)) THEN
+      capacity = three_parameter_connection(temperature, liquid_fraction, &
+        Dcapacity_solid, Dcapacity_liquid, capacity_fusion)
+    ELSE
+      capacity = three_parameter_connection(temperature, liquid_fraction, &
+        Dcapacity_solid, Dcapacity_liquid, capacity_fusion, .TRUE.)
+    END IF
   END FUNCTION capacity
 
   FUNCTION temperature (enthalpy)
     IMPLICIT NONE
     REAL (pr), INTENT(IN) :: enthalpy(:)
     REAL (pr) :: temperature(SIZE(enthalpy))
-    IF (capacity_der.LE.1e-6) THEN
-        temperature = enthalpy_fusion(enthalpy)
-    ELSE
-        !temperature = enthalpy_fusion(enthalpy)
-        temperature = (SQRT(1.0_pr + 4*capacity_der*enthalpy_fusion(enthalpy)) - 1.0_pr)/2/capacity_der
-    END IF
+    REAL (pr) :: h1(SIZE(enthalpy)), h2(SIZE(enthalpy)), liquid_fraction(SIZE(enthalpy))
+    liquid_fraction = liquid_fraction(enthalpy)
+    h2 = Dcapacity_solid*(1.0_pr - liquid_fraction) + Dcapacity_liquid*liquid_fraction
+    WHERE (h2.GE.1e-6)
+      h1 = 1.0_pr*(1.0_pr - liquid_fraction) + &
+        (1.0_pr + Dcapacity_solid - Dcapacity_liquid + capacity_fusion)*liquid_fraction
+      temperature = (SQRT(h1**2 + 2*h2*enthalpy_fusion(enthalpy)) - h1)/h2
+    ELSEWHERE
+      !seems limit open also right that
+      temperature = enthalpy_fusion(enthalpy)
+    END WHERE
   END FUNCTION temperature
 
+  FUNCTION temperature_derivative (enthalpy_prev)
+    IMPLICIT NONE
+    REAL (pr), INTENT(IN) ::  enthalpy_prev(:)
+    REAL (pr) :: temperature_derivative(SIZE(enthalpy_prev))
+    REAL (pr) :: dh2(SIZE(enthalpy_prev)), dh1(SIZE(enthalpy_prev)), h1(SIZE(enthalpy_prev)), h2(SIZE(enthalpy_prev))
+
+    h1 = 1.0_pr + (Dcapacity_solid - Dcapacity_liquid + capacity_fusion)*liquid_fraction(enthalpy_prev)
+    h2 = Dcapacity_solid*(1.0_pr - liquid_fraction(enthalpy_prev)) + Dcapacity_liquid*liquid_fraction(enthalpy_prev)
+    dh1 = (Dcapacity_solid - Dcapacity_liquid + capacity_fusion)*liquid_fraction(enthalpy_prev, .TRUE.)
+    dh2 = (Dcapacity_liquid - Dcapacity_solid)*liquid_fraction(enthalpy_prev, .TRUE.)
+    temperature_derivative = (-temperature(enthalpy_prev)*dh2/h2 - dh1/h2 + &
+      (enthalpy_fusion(enthalpy_prev)*dh2 + h1*dh1 + h2*enthalpy_fusion(enthalpy_prev, .TRUE.)))/ &
+      (temperature(enthalpy_prev) + h1)
```

Comment:
Deprecated and wrong version. Implement (28) and (29). And examine them thoroughly before!